src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+If an algorithm throws an exception, it means returning a Completion Record with [[Type]] throw whose [[Value]] is the exception object. We'll ignore the break, continue and return types for now.
+
+_ReturnIfAbrupt(argument)_ means taking the following steps:
+
+> 1. If argument is abrupt, return argument
+> 2. Set argument to argument.\[\[Value\]\]
+
+That is, we inspect a Completion Record; if it's an abrupt completion, we return immediately. Otherwise, we extract the value from the Completion Record.
+
+[Spec: ReturnIfAbrupt](https://tc39.es/ecma262/#sec-returnifabrupt)
+
+_ReturnIfAbrupt_ might look like a function call, but it's not. It causes the function where _ReturnIfAbrupt()_ occurs to return, not the "ReturnIfAbrupt" function itself. It behaves more like a macro in C-like languages.
+
+_ReturnIfAbrupt_ can be used like this:
+
+> let obj be Foo()
+> // obj is a Completion Record
+> ReturnIfAbrupt(obj)
+> // If we're still here, obj is now the value extracted from the Completion Record
+> Bar(obj)
+
+And now the question mark comes into play: ? Foo() is equivalent to _ReturnIfAbrupt(Foo())_.
+
+Similarly, Let val be ! Foo() is equivalent to:
+
+> let val be Foo()
+> Assert: val is not an abrupt completion
+> Set val to val.\[\[Value\]\].
+
+[Spec: ReturnIfAbrupt shorthands](https://tc39.es/ecma262/#sec-returnifabrupt-shorthands)
+
+Using this knowledge, we can rewrite _Object.prototype.hasOwnProperty_ like this:
+
+> Object.prototype.hasOwnProperty(P)
+>
+> 1. Let P be ToPropertyKey(V).
+> 2. If P is an abrupt completion, return P
+> 3. Set P to P.[[Value]]
+> 4. Let O be ToObject(this value).
+> 5. If O is an abrupt completion, return O
+> 6. Set O to O.[[Value]]
+> 7. Let temp be HasOwnProperty(O, P).
+> 8. If temp is an abrupt completion, return temp
+> 9. Let temp be temp.\[\[Value\]\]
+> 9. Return NormalCompletion(temp)
+
+and _HasOwnProperty_ like this:
+
+> HasOwnProperty(O, P)
+>
+> 1. Assert: Type(O) is Object.
+> 2. Assert: IsPropertyKey(P) is true.
+> 3. Let desc be O.\[\[GetOwnProperty\]\](P).
+> 4. If desc is an abrupt completion, return desc
+> 5. Set desc to desc.[[Value]]
+> 6. If desc is undefined, return NormalCompletion(false).
+> 7. Return NormalCompletion(true).
+
+We can also rewrite the _\[\[GetOwnProperty\]\]_ internal method without the exclamation mark like this:
+
+> O.\[\[GetOwnProperty\]\]
+>
+> 1. Let temp be OrdinaryGetOwnProperty(O, P)
+> 2. Assert: temp is not an abrupt completion
+> 3. Let temp be temp.\[\[Value\]\]
+> 4. Return NormalCompletion(temp)
+
+Here we assume that _temp_ is a brand new temporary variable which doesn't collide with anything else.
+
+We've also used the knowledge that when a return statement returns something else than a Completion Record, it's implicitly wrapped inside a NormalCompletion.
```

Comment:
Sorry to harp on this point. I'd add a note here that even though the implicit rule is in place, the spec itself isn't always consistent. Often predicate-like AOs are called and their bare return values are used without unwrapping via `?` and `!`. A note to the reader that this is usually clear from context, though it is being worked on upstream.

Reply:
Added a note (not here but above where this "implicit completion" thing is introduced for the first time).